x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/decider/SearchableSnapshotRepositoryExistsAllocationDecider.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.searchablesnapshots.allocation.decider;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.RepositoriesMetadata;
+import org.elasticsearch.cluster.metadata.RepositoryMetadata;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDecider;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsConstants;
+
+import java.util.List;
+
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_REPOSITORY_NAME_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_REPOSITORY_UUID_SETTING;
+
+public class SearchableSnapshotRepositoryExistsAllocationDecider extends AllocationDecider {
+
+    private static final String NAME = "searchable_snapshot_repository_exists";
+
+    private static final Decision YES_INAPPLICABLE = Decision.single(
+        Decision.Type.YES,
+        NAME,
+        "this decider only applies to indices backed by searchable snapshots"
+    );
+
+    private static final Decision YES_REPOSITORY_EXISTS = Decision.single(
+        Decision.Type.YES,
+        NAME,
+        "the repository containing the data for this index exists"
+    );
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return allowAllocation(allocation.metadata().getIndexSafe(shardRouting.index()), allocation);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocation) {
+        return allowAllocation(allocation.metadata().getIndexSafe(shardRouting.index()), allocation);
+    }
+
+    @Override
+    public Decision canAllocate(IndexMetadata indexMetadata, RoutingNode node, RoutingAllocation allocation) {
+        return allowAllocation(indexMetadata, allocation);
+    }
+
+    private static Decision allowAllocation(IndexMetadata indexMetadata, RoutingAllocation allocation) {
+        final Settings settings = indexMetadata.getSettings();
+        if (SearchableSnapshotsConstants.isSearchableSnapshotStore(settings)) {
+
+            final RepositoriesMetadata repositoriesMetadata = allocation.metadata().custom(RepositoriesMetadata.TYPE);
```

Comment:
If we restore a searchable snapshot from another repo and the searchable snapshot repo is unavailable, I think this means that the restore operation will not complete until the repo becomes available. I wonder if it was more intuitive to have the restore fail? I.e., to allow the allocation always if there is a restore UUID other than the "no api" UUID.

Reply:
I believe if a shard cannot be immediately allocated then it is recorded as a restore failure (and therefore complete) which lets the restore complete:

https://github.com/elastic/elasticsearch/blob/e491cdd23f92626a580ff51007cf479d490ec35e/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L673-L678

However I haven't checked this yet. Was this what you meant?

Reply:
Yes, thanks for clarifying this.